File: apps/web/src/lib/sentry.ts
import { useEffect } from "react";
import * as Sentry from "@sentry/react";
import {
  createRoutesFromChildren,
  matchRoutes,
  useLocation,
  useNavigationType,
} from "react-router-dom";
import type { WebConfig } from "../modules/config/config";

const REDACTED = "[Filtered]";

const SENSITIVE_KEY_PARTS = [
  "authorization",
  "cookie",
  "token",
  "secret",
  "signature",
  "apikey",
  "signedurl",
  "downloadurl",
  "uploadurl",
  "sampleurl",
  "manuscripttext",
  "documenttext",
  "documentchunk",
  "chunktext",
  "rawtext",
  "fulltext",
  "extractedtext",
  "contact",
  "recipient",
  "email",
  "phone",
  "card",
  "cvv",
  "pan",
];

const SENSITIVE_QUERY_KEY_PARTS = [
  "token",
  "signature",
  "credential",
  "key",
  "secret",
  "x-goog-signature",
  "x-goog-credential",
  "x-goog-security-token",
];

let initialized = false;

export const SentryErrorBoundary = Sentry.ErrorBoundary;
export const SENTRY_REDACTED_VALUE = REDACTED;

export function initializeSentry(config: WebConfig): void {
  if (!config.sentryDsn || initialized) {
    return;
  }

  Sentry.init({
    dsn: config.sentryDsn,
    environment: config.sentryEnvironment,
    release: config.sentryRelease,
    sendDefaultPii: false,
    tracesSampleRate: config.sentryTracesSampleRate,
    tracePropagationTargets: [config.apiBaseUrl],
    integrations: [
      Sentry.reactRouterV7BrowserTracingIntegration({
        useEffect,
        useLocation,
        useNavigationType,
        createRoutesFromChildren,
        matchRoutes,
      }),
    ],
    beforeSend(event) {
      return scrubSentryEvent(event);
    },
  });

  Sentry.setTag("service", "web");
  initialized = true;
}

export function scrubSentryEvent<T>(event: T): T {
  return scrubValue(event, new WeakSet<object>()) as T;
}

function scrubValue(value: unknown, seen: WeakSet<object>): unknown {
  if (typeof value === "string") {
    return scrubString(value);
  }

  if (Array.isArray(value)) {
    return value.map((item) => scrubValue(item, seen));
  }

  if (typeof value !== "object" || value === null) {
    return value;
  }

  if (seen.has(value)) {
    return REDACTED;
  }

  seen.add(value);

  return Object.fromEntries(
    Object.entries(value as Record<string, unknown>).map(([key, nested]) => [
      key,
      shouldRedactKey(key) ? REDACTED : scrubValue(nested, seen),
    ]),
  );
}

function shouldRedactKey(key: string): boolean {
  const normalized = key.toLowerCase().replace(/[^a-z0-9]/g, "");
  return SENSITIVE_KEY_PARTS.some((part) => normalized.includes(part));
}

function scrubString(value: string): string {
  if (
    value.includes("X-Goog-Signature") ||
    value.includes("X-Goog-Credential")
  ) {
    return REDACTED;
  }

  return scrubUrl(value);
}

function scrubUrl(value: string): string {
  try {
    const url = new URL(value);
    let changed = false;

    for (const key of Array.from(url.searchParams.keys())) {
      const normalized = key.toLowerCase();
      if (SENSITIVE_QUERY_KEY_PARTS.some((part) => normalized.includes(part))) {
        url.searchParams.set(key, REDACTED);
        changed = true;
      }
    }

    return changed ? url.toString() : value;
  } catch {
    return value;
  }
}

File: apps/ai-service/app/modules/sentry.py
from __future__ import annotations

from typing import Any, cast
from urllib.parse import parse_qsl, urlencode, urlsplit, urlunsplit

import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.httpx import HttpxIntegration
from sentry_sdk.types import Event, Hint

from app.modules.config import AiServiceConfig

REDACTED = "[Filtered]"

SENSITIVE_KEY_PARTS = (
    "authorization",
    "cookie",
    "token",
    "secret",
    "signature",
    "apikey",
    "servicerole",
    "signedurl",
    "downloadurl",
    "uploadurl",
    "sampleurl",
    "manuscripttext",
    "documenttext",
    "documentchunk",
    "chunktext",
    "rawtext",
    "fulltext",
    "extractedtext",
    "contact",
    "recipient",
    "email",
    "phone",
    "card",
    "cvv",
    "pan",
)

SENSITIVE_QUERY_KEY_PARTS = (
    "token",
    "signature",
    "credential",
    "key",
    "secret",
    "x-goog-signature",
    "x-goog-credential",
    "x-goog-security-token",
)


def initialize_sentry(config: AiServiceConfig) -> None:
    if not config.sentry_dsn:
        return

    sentry_sdk.init(
        dsn=config.sentry_dsn,
        environment=config.sentry_environment,
        release=config.sentry_release,
        send_default_pii=False,
        traces_sample_rate=config.sentry_traces_sample_rate,
        integrations=[FastApiIntegration(), HttpxIntegration()],
        before_send=scrub_sentry_event,
    )
    sentry_sdk.set_tag("service", "ai-service")


def scrub_sentry_event(event: Event, hint: Hint | None = None) -> Event:
    _ = hint
    return cast(Event, _scrub_value(event))


def _scrub_value(value: Any) -> Any:
    if isinstance(value, str):
        return _scrub_string(value)

    if isinstance(value, list):
        return [_scrub_value(item) for item in value]

    if isinstance(value, dict):
        return {
            key: REDACTED if _should_redact_key(str(key)) else _scrub_value(nested)
            for key, nested in value.items()
        }

    return value


def _should_redact_key(key: str) -> bool:
    normalized = "".join(character for character in key.lower() if character.isalnum())
    return any(part in normalized for part in SENSITIVE_KEY_PARTS)


def _scrub_string(value: str) -> str:
    if "X-Goog-Signature" in value or "X-Goog-Credential" in value:
        return REDACTED

    if "service_role" in value.lower() or "sentry_auth_token" in value.lower():
        return REDACTED

    return _scrub_url(value)


def _scrub_url(value: str) -> str:
    parsed = urlsplit(value)
    if not parsed.scheme or not parsed.netloc:
        return value

    query = parse_qsl(parsed.query, keep_blank_values=True)
    changed = False
    scrubbed_query: list[tuple[str, str]] = []

    for key, nested_value in query:
        normalized = key.lower()
        if any(part in normalized for part in SENSITIVE_QUERY_KEY_PARTS):
            scrubbed_query.append((key, REDACTED))
            changed = True
        else:
            scrubbed_query.append((key, nested_value))

    if not changed:
        return value

    return urlunsplit(
        (
            parsed.scheme,
            parsed.netloc,
            parsed.path,
            urlencode(scrubbed_query),
            parsed.fragment,
        )
    )
